sdk/inc/azure/iot/az_iot_provisioning_client.h

```diff
@@ -423,6 +423,45 @@ AZ_NODISCARD az_result az_iot_provisioning_client_query_status_get_publish_topic
     size_t mqtt_topic_size,
     size_t* out_mqtt_topic_length);
 
+/**
+ * @brief Azure IoT Provisioning Client options for az_iot_provisioning_client_get_request_payload.
+ *        Not currently used.  Reserved for future use.
+ *
+ */
+typedef void az_iot_provisioning_client_payload_options;
+
+/**
+ * @brief Builds the optional payload for a provisioning request.
+ * @remark Use this API to build an MQTT payload during registration.
+ *         This call is optional for most scenarios.  Some service
+ *         applications may require #custom_payload_property specified during
+ *         registration to take additional decisions during provisioning time.
+ *         For example, if you need to register an IoT Plug and Play device you must
+ *         specify its model_id with this API via the #custom_payload_property
+ *         "{\"modelId\":\"" your_model_id "\"}".
+ *
+ * @param[in] client The #az_iot_provisioning_client to use for this call.
+ * @param[in] custom_payload_property __[nullable]__ Custom JSON to be added to this payload.
+ * Can be `NULL`.
+ * @param[in] reserved __[nullable]__ Reserved field for future options to this function.
+ * @param[out] mqtt_payload A buffer with sufficient capacity to hold the MQTT payload.
+ * @param[in] mqtt_payload_size The size, in bytes of \p mqtt_payload.
+ * @param[out] out_mqtt_payload_length __[nullable]__ Contains the string length, in bytes, of \p
+ * mqtt_payload. Can be `NULL`.
+ * @pre \p client must not be `NULL`.
+ * @pre \p reserved must be `NULL`
+ * @pre \p mqtt_payload must not be `NULL`.
+ * @pre \p mqtt_payload_size must be greater than 0.
+ * @return An #az_result value indicating the result of the operation.
```

Comment:
Worth documenting some of the common failure conditions/return codes?

Reply:
@ahsonkhan - Great question.  I can do that after... my question to you is to double check you're OK with the word "some".

In general when I see API documentation with error codes, I always assume that the API could return other stuff, too, so I'm forgiving as the caller.  I think that's what most developers would do.

I'm asking here because this API's errors are ultimately going to be coming from az_json_writer, which could in theory add new errors in future that just realistically we're not going to update in the az_prov_ space.  

Also most functions in this header don't have retval.  Once we're on same page about "some", I'll open a GitHub issue to track adding them.

Reply:
Yep, I totally agree.

Trying to list and solve the "transitive closure" of all error codes possible from all callers within the call tree of a particular function could become a daunting task and wouldn't scale well from a docs perspective (and as impl changes/evolves, an untenable maintenance burden).

We had loosely discussed this in the past (at least within the core team), and since we were generally at the bottom, listing more was a bit easier.

This becomes more of an art than a science, and the implementer of the method can dictate what are "some" of the more relevant or most commonly occurring return codes the user cares about that, specific to this, and would be useful for the caller to know. If the dev doesn't have a good sense of which codes might be useful in the caller scenario, maybe "tree-pruning" to just 1 level down the call stack for certain impl details is a reasonable heuristic if we are flying blind (while keeping an eye on the list not growing too large/becoming too noisy).

Reply:
So, yes, this required a bit of a judgement call, and we should let the caller scenario drive the decision of what error codes to highlight in docs. One question that might be useful to ask is: What error cases do I expect most callers of my function to handle/write code against?

Reply:
If they don't get OK, the error overwhelmingly likely to cause this to go wrong would be a `AZ_ERROR_NOT_ENOUGH_SPACE`.  I'll document that.

Trying to fix up other places in this header that need it is well beyond scope of this PR.  Opened #1878 as our IOU so we don't lose this.